### ISSUE TITLE
Build the V1 evaluation harness

### ISSUE THREAD

**Issue by mark-torres10** [METResearchGroup/moral_outrage_classifier#4]
# Issue Description

Now that we've got the [Perspective API to work](https://github.com/METResearchGroup/moral_outrage_classifier/issues/2), let's create an evaluation harness.

Ideally this would be called from the command line and save the results to a .csv file:

The basic shape would look something like:

```markdown
for each row:
  for each predictor (your model, GPT prompt, Perspective):
    call predictor(text) -> pred_label or pred_score
    store row in results
```

The CLI flags would be something like:

```bash
--input_path data/reddit_labeled.csv
--model your_api,gpt,perspective (or default all)
--output_path results/reddit_run_20260327.csv
--max-rows 500 # We use --max-rows for smoke testing. We don't want to spend $50 on API calls just to find out our CSV export has a typo in the header. Always run 5 rows before running 5,000
```

In the saved .csv file, we'd want something like this per-row:

```plaintext
- id: some identifier for the text
- dataset: the path to the .csv file
- gold_label: 0/1
- model_name: model name
- pred_label: predicted label
- pred_score: predicted score (float, between 0 and 1). Only applicable for Perspective API, else will be None for any LLMs (those will give binary scores, not probability floats).
- is_correct: 0/1 (binary)
```

Each row should be unique on the combination of (id, model_name).

Right now, let's just set this up to work on the Perspective API.

## Expected file structure

```bash
- evaluation/
    - dataloader.py # implements DataLoader
    - run_evaluation_harness.py
    - examples_test.py # example test, to show how it's run.
```

### Step 1: Prepare a dataloader

We'll first prepare a [dataloader](https://www.kaggle.com/discussions/general/406267). A dataloader in machine learning is a utility that efficiently loads and preprocesses data from a dataset and prepares it in small chunks (mini-batches) for a model during training or inference. Note: if you look up dataloader online, you'll see a `torch.utils.data.DataLoader` mentioned. This is overkill for our use cases, we'll want to create our own version.

```python
class DataLoader:
      def __init__(self, input_path: str, batch_size: int):
           self.data: list[tuple[str, int]] = []
           self.batch_size = batch_size

      def load_data(self):
           # deal with all the steps related to loading the data. Should save the resulting output to `self.data` in (input, output) pairs, where input = the text string, output = label (0/1) for moral outrage.
           # also should manage filtering out "already-processed" IDs. This way, EvaluationHarness can assume that loaded data is already filtered out.
           pass

    def filter_already_processed_records(self):
          pass
  
     def __iter__(self) -> list[tuple[str, int]]:
          # review "dunder methods" and "__iter__" on how to define iterators. This method should iterate the self.data to get a batch of  [(input, output)] pairs during iteration. For example, if batch size = 10, we should have [(x_1, y_1), ..., (x_10, y_10)]. Make sure to make this also robust if the sample size is not a multiple of the batch size (e.g., you have 9 training samples but a batch size of 10, or if you have 21 training samples but a batch size of 10). Also look up generators in Python and the difference between return vs. yield.
```

This should be called like this:

```
input_path = "data/reddit.csv"
batch_size = 10 # let's default to 10 for now
dataloader = DataLoader(input_path=input_path, batch_size=batch_size)

for batch in dataloader:
    training_samples = []
    labels = []
    for sample in batch:
         training_samples.append(sample[0])
         labels.append(sample[1])
   ...
```

Please write unit tests for this file as well, in tests/evaluation/test_dataloader.py. Also install `pytest` and create the CI file that will run the pytest (see https://github.com/METResearchGroup/social_agent_simulation_platform/blob/main/.github/workflows/ci.yml for an example of this).

The DataLoader should be a lazy generator. It should not load the entire 50,000-row CSV into memory at once. It should stream rows so we can eventually handle datasets larger than our RAM.

### Step 2: Create evaluation harness

Now that we have this, we can create the evaluation harness.

```python
from models.perspective_api.model import PerspectiveAPIModel

VALID_MODELS = ["perspective_api"]

class EvaluationHarness:
      def __init__(
          self,
          input_path: str,
          output_path: dir,
          batch_size: int,
          models: list[str]
     ):
           self.dataloader = DataLoader(input_path=input_path, batch_size=batch_size)
           
      @classmethod # look up what a classmethod is and why we would use it here.
      def validate_models(cls, models: list[str]):
             if len(models) == 0 or any(model not in VALID_MODELS for model in models):
             raise ValueError(f"Please provide a valid model. {models} is invalid")

     def load_data(self):
           self.dataloader.load_data()

      def run_evaluation():
           # here, I leave implementation up to you. Please keep in mind the design questions below. You should run verification and save results.
           # 1. Initialize the specific model class based on model_name (or perhaps initialize multiple models here? Think about the pros/cons based on your desired design, see below)
           # 2. Check for existing results in output_dir (for idempotency)
           # 3. Iterate through self.dataloader
           # 4. Save results incrementally
           pass

     def display_results():
          # display the results as a nice `rich` table.
```

Ideally, we want some variation of the public interface to look like this (imagine putting this in the `if __name__ == __main__` block:

```python
# should grab these from the CLI.
input_path = ""
output_path = ""
batch_size = 10
models = ["perspective_api"]

eval_harness = EvaluationHarness(
    input_path=input_path,
    output_path=output_path,
    batch_size=batch_size,
    models=models
)
eval_harness.load_data()
eval_harness.run_evaluation()
eval_harness.display_results()
```

The details are intentionally left open-ended. I'd suggest starting at the intended public interface and doing each chunk step by step.
      
## Design choices to consider during implementation

Please address these in your PR description, as these are important design choices for making this evaluation harness.

- Should these all be run one-at-a-time? Or in parallel? If parallel, parallel at what unit? (run all the labels for a single text all at once? Run all the labels for a single classifier and then move on to the next classifier? Something in between?)
- What happens if one model call fails? Should that screw up everything else?
- What are the [idempotency guarantees](https://medium.com/cache-me-out/understanding-idempotency-68a50a837fc1)? For example, what happens if we label a post twice? We ideally don't want to label something multiple times (that just wastes money, for example). How do we protect against this? (hint: explore different caching principles as well as checking if a label exists before you label it. The easiest solution is just loading existing labels and then filtering for those). We really want to make sure we don't double-classify posts unnecessarily.
- How are you saving the data? As .json? As .csv? And how are you saving the metadata of the run that was used to generate the results? One example setup is:

```python
<output_path>/<timestamp>/
     output.csv
     metadata.json # please put (1) the git commit hash, (2) timestamp, and (3) CLI args.
```

For CLI args, please use the `Typer` library.

- How often are you saving the results? Only when the entire batch is done? One result at a time? Only when the entire training dataset is done? The harness must persist results to disk incrementally (e.g., after every batch) so that progress isn't lost on a network failure.

These are important to figure out now because LLM inference is much more finicky than the Perspective API. You'll need to figure out how to manage cases where the model gives you nonsense results, you get HTTP errors, etc.

## What defines "done"?

- Implementation of above.
- Benchmarks related to runtime (like in previous PRs).
- Answers to the provided design questions.
